### ISSUE TITLE
logind: external monitor detection broken on wayland

### ISSUE THREAD

**Issue by yorickvP** [systemd/systemd#12001]
**systemd version the issue has been seen with**
> systemd 239

**Used distribution**
 > NixOS + sway 1.0

**Expected behaviour you didn't see**
 > when putting the external screens in dpms and closing the lid, laptop should not sleep

**Unexpected behaviour you saw**
 > laptop sleeps

**Steps to reproduce the problem**
```
# close laptop lid, with external screens connected
$ swayidle timeout 300 'swaymsg "output * dpms off"; swaylock' resume 'swaymsg "output * dpms on"'
$ pkill -USR1 swayidle
# system suspends
```

On wayland systems, there is no way to distinguish outputs in DPMS from disabled outputs (they are the same from a kernel point of view).

**Comment by emersion** [comment]
Here are a few more details:

The [DRM docs](https://01.org/linuxgraphics/gfx-docs/drm/gpu/drm-kms.html#standard-connector-properties) say there are two properties to turn off outputs: "active" and "DPMS". However, "DPMS" is legacy and cannot be changed via the atomic interface. So instead sway is using the "active" property, which makes the output appear as disabled via sysfs.

I suspect this bug appears on more Wayland compositors, and will eventually be the case for everyone (including xorg-server) as they migrate from the legacy interface to the atomic interface.

There's no way to know whether an output has been disabled or put into power-save mode with the new DRM API.

**Comment by poettering** [comment]
So, is there a kernel API that allows us to distinguish "port is off because no hw connected" and "port is off because i told my display server to"? if not I fear we can't really do anything about this.

**Comment by yorickvP** [comment]
Yes, hw connected is at `/status`, but systemd checks `/enabled` as well as `/status`.

Solutions I can think of are handling it in userspace (inhibiting handle-lid-switch check), or removing the `enabled` check in systemd.

**Comment by nolanl** [comment]
> I suspect this bug appears on more Wayland compositors, and will eventually be the case for everyone (including xorg-server) as they migrate from the legacy interface to the atomic interface.

I just upgraded from Debian Stretch to Buster, and am now hitting this with xorg-server.  Buster has systemd 241.

Reproducing it is as simple as running "xset dpms force off" "xset dpms force suspend" or "xset dpms force standby" with the lid closed and an external monitor attached.

**Comment by nolanl** [comment]
I suspect the reason this bug isn't causing more widespread issues is that for GNOME users, gsd-power appears to inhibit systemd's lid handling, and does everything itself in userspace.

**Comment by mato** [comment]
Apologies if this is a different bug, but I'm in a similar situation to that of @nolanl. I'm the author of `xscreensaver-systemd`, which adds support for Inhibitor Locks to XScreenSaver, and is integrated in recent releases thereof. After upgrading to Debian Buster, I started seeing issues with _waking up from suspend_ (via a previous manual `systemctl suspend`) and an external monitor connected.

I originally put this down to some bug on the xscreensaver side of things, since we also got reports of broken suspend/resume behaviour from other users. However, after finding this bug and trying @nolanl's method of reproduction, I'm inclined to think systemd may be at fault.

In an attempt to reproduce with as few moving pieces as possible, I've tried the following:

- plain Xorg, `xserver-xorg-video-intel` driver, custom latest 4.19 LTS kernel.
- no desktop environment or xscreensaver running.
- lid closed, external monitor connected.

On running `xset dpms force off`: the monitor goes into power saving mode and the laptop does _not_ suspend. I.e. the behaviour is correct.

However, when I press a key / move the mouse to switch the monitor on what seems to happen most times is that:

1. Monitor starts to power up (observed by the backlight switching on).
2. Laptop immediately goes into suspend (observed by both the power LED and system logs, and monitor again reporting "entering power save mode").
3. On moving the mouse / pressing keys a _second time_, the laptop wakes up normally.

(2) above matches what we see in reports from some xscreensaver-systemd users logs even with _no external monitor_ connected. So it seems like something is racy on the _wakeup_ path.

@poettering Any ideas? Let me know what debug knobs I should trigger in systemd/systemd-logind to provide logs.

**Comment by keszybz** [comment]
@mato, your issue sounds like #16045. Please check if the PR linked there helps.

**Comment by keszybz** [comment]
As for the original issue reported by @yorickvP: I think we need a userspace solution to this. Essentially, when we disable a display because we **temporarily** don't need the screen, it doesn't really matter what mechanism is used. For the user, a (non-power-saving) screensaver, dpms, or setting the display to disabled using the atomic interface is all the same: they don't see anything on the screen, but the machine is still being otherwise used, and we should not suspend or take another similar action. Using the atomic disable option is actually pretty good for this, because it should result in maximum power savings. But the kernel doesn't know if the display is taking a nap or if it is being permanently disabled because the user doesn't need that screen. My conclusion is that it's the userspace that needs to prevent suspend if it is temporarily disabling a display. Thus, I think `sway` should somehow inform logind that it's only temporarily disabling the display and it should be counted as "connected". Not sure though what the best way to transmit this information would be.

**Comment by emersion** [comment]
... Or systemd shouldn't try to guess when to sleep based on heuristics like "is the screen DPMS'ed".

**Comment by keszybz** [comment]
> ... Or systemd shouldn't try to guess when to sleep based on heuristics like "is the screen DPMS'ed".

Pfff, it doesn't are about DPMS. I cares about external monitors being connected.

**Comment by emersion** [comment]
Well, it seems like DPMS'ing external screens results in systemd putting the system into sleep, which is incorrect. So I'd say systemd pretty much cares about DPMS.

**Comment by keszybz** [comment]
Based on the description in the bug report, DPMS is implemented as deactivating of the monitor. No need to repeat that it's DPMS — the effect is similar, but it is not DPMS.

**Comment by yorickvP** [comment]
Systemd cares if external monitors are actually used, as well as connected. It seems that with recent DRM, it's no longer possible to distinguish unused monitors from temporarily disabled monitors with just that API.
To mark this issue as resolved, it looks like either systemd's behavior needs to be reverted to only looking at connected monitors, or anyone temporarily disabling monitors should take manual control of logind's behavior.

**Comment by emersion** [comment]
A workaround is to set `HandleLidSwitch=ignore` in `logind.conf` (and configure a lid switch binding in Sway if desirable).

**Comment by RyanGibb** [comment]
> A workaround is to set `HandleLidSwitch=ignore` in `logind.conf` (and configure a lid switch binding in Sway if desirable).

This would mean a lid switch outside a display server wouldn't trigger, e.g., a suspend.

A workaround I found preferable was leaving `HandleLidSwitch=suspend`, and leaving the lid open when docked but disabling the internal screen.
